papermill/preprocess.py

```diff
@@ -202,3 +194,81 @@ def papermill_preprocess(self, nb, resources):
                     cell.metadata['papermill']['status'] = COMPLETED
                     future.result()
         return nb, resources
+        
+    def run_cell(self, cell, cell_index=0):
+        msg_id = self.kc.execute(cell.source)
+        self.log.debug("Executing cell:\n%s", cell.source)
+        outs = cell.outputs = []
+
+        while True:
+            try:
+                # We are not waiting for execute_reply, so all output
+                # will not be waiting for us. This may produce currently unknown issues.
+                msg = self.kc.iopub_channel.get_msg(timeout=None)
+            except Empty:
+                self.log.warning("Timeout waiting for IOPub output")
+                if self.raise_on_iopub_timeout:
+                    raise RuntimeError("Timeout waiting for IOPub output")
+                else:
+                    break
+            if msg['parent_header'].get('msg_id') != msg_id:
+                # not an output from our execution
+                continue
+
+            msg_type = msg['msg_type']
+            self.log.debug("output: %s", msg_type)
+            content = msg['content']
+
+            # set the prompt number for the input and the output
+            if 'execution_count' in content:
+                cell['execution_count'] = content['execution_count']
+
+            if msg_type == 'status':
+                if content['execution_state'] == 'idle':
+                    break
+                else:
+                    continue
+            elif msg_type == 'execute_input':
+                sys.stdout.write('Executing Cell {:-<40}\n'.format(content.get("execution_count", "*")))
```

Comment:
Need an `if self.log_output: ` here so we only log this when we have set the --log-outputs option.

Reply:
This should also use stderr instead of stdout as it's a message instead of data (for good unix-like command support)